### DISCUSSION TITLE
[mini.completion] Confusion between `completefunc` and `omnifunc`

### DISCUSSION THREAD

**Discussion by pkazmier** [nvim-mini/mini.nvim#811] [Q&A] [answered]
Embarrassingly, I've spent the last 3 hours trying to figure out why completion behaves differently when the `source_func` is `"completefunc"` vs `"omnifunc"`.  For example, if my buffer contains:

```lua
local H ={}
H.test1 = function() end
H.test2 = function() end
H.test3 = function() end
```

When `source_func = "completefunc"`:

1. `Go`, end of file, open line
2. `H`, popup opens with "H variable" (expected)
3. `.`, popup opens with "test1() Function", "test2() Function", and "test3() Function" (expected)

When `source_func = "omnifunc"`:

1. `Go`, end of file, open line
2. `H`, popup opens with "H variable" (expected)
3. `.`, popup dismissed (not expected)

I've added print statements all over `mini.completion` and it appears that the event `CompleteDonePre` is fired when I press `.` with `"completefunc"`, but it is not fired with `"omnifunc"`. When `CompleteDonePre` is fired, it results in `H.stop_completion` being passed `nil`, which resets the source.

And to continue the `"omnifunc"` above:

4. `t`, popup opens showing "test1", "test2", "test3". These come from the fallback, not LSP, because the source was not reset.

I suppose my expectation of how it should work is probably due to some fundamental misunderstanding about vim's completion function vs omni function. I tried reading help pages, but have not found the answer. It's not for the lack of trying. Can you help elaborate for me?

I prefer the behavior when the source in `"completefunc"`, but I only swapped it because of the note in `mini.completion` about the better way of hooking this up is to do so in the LSP `on_attach` function. I wish I never read that paragraph now as I could have saved my entire morning 😄

**Comment by pkazmier** [comment]
Here is a minimal configuration one can test with:

```lua
-- Clone 'mini.nvim' manually in a way that it gets managed by 'mini.deps'
local path_package = vim.fn.stdpath("data") .. "/site/"
local mini_path = path_package .. "pack/deps/start/mini.nvim"
if not vim.loop.fs_stat(mini_path) then
  vim.cmd('echo "Installing `mini.nvim`" | redraw')
  local clone_cmd = {
    "git",
    "clone",
    "--filter=blob:none",
    "https://github.com/echasnovski/mini.nvim",
    mini_path,
  }
  vim.fn.system(clone_cmd)
  vim.cmd("packadd mini.nvim | helptags ALL")
  vim.cmd('echo "Installed `mini.nvim`" | redraw')
end

-- Set up 'mini.deps' (customize to your liking)
require("mini.deps").setup({ path = { package = path_package } })
MiniDeps.add("williamboman/mason.nvim")
MiniDeps.add("williamboman/mason-lspconfig.nvim")
MiniDeps.add("neovim/nvim-lspconfig")

require("mini.completion").setup({
  lsp_completion = {
    source_func = "omnifunc",
    auto_setup = true,
  },
})
require("mason").setup()
require("mason-lspconfig").setup({
        ensure_installed = { "lua_ls" }
})
require("lspconfig").lua_ls.setup({
  on_attach = function(client, bufnr)
    -- vim.bo[bufnr].omnifunc = "v:lua.MiniCompletion.completefunc_lsp"
  end,
})
```

**Comment by echasnovski** [comment]
Thanks for the detailed description with explicit steps and setup!

> Embarrassingly, I've spent the last 3 hours trying to figure out why completion behaves differently when the `source_func` is `"completefunc"` vs `"omnifunc"`.

I can reproduce even with by slightly tweaking my own setup. Plan to take a look closer in the near future.

> I suppose my expectation of how it should work is probably due to some fundamental misunderstanding about vim's completion function vs omni function. I tried reading help pages, but have not found the answer. It's not for the lack of trying. Can you help elaborate for me?

It is surely can be an issue in 'mini.completion'. That said, I also am not aware of the obvious difference between how 'completefunc' and 'omnifunc' operates.

**Comment by echasnovski** [comment, marked as answer]
This should be solved on latest `main`.
